src/otx/engine/engine.py

```diff
@@ -0,0 +1,407 @@
+"""Module for OTX engine components."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Iterable
+
+import torch
+from lightning import Trainer, seed_everything
+
+from otx.core.config.device import DeviceConfig
+from otx.core.data.module import OTXDataModule
+from otx.core.model.entity.base import OTXModel
+from otx.core.model.module.base import OTXLitModule
+from otx.core.types.device import DeviceType
+from otx.core.types.task import OTXTaskType
+from otx.core.utils.cache import TrainerArgumentsCache
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from lightning import Callback
+    from lightning.pytorch.cli import LRSchedulerCallable, OptimizerCallable
+    from lightning.pytorch.loggers import Logger
+    from lightning.pytorch.utilities.types import EVAL_DATALOADERS
+    from pytorch_lightning.trainer.connectors.accelerator_connector import _PRECISION_INPUT
+
+
+LITMODULE_PER_TASK = {
+    OTXTaskType.MULTI_CLASS_CLS: "otx.core.model.module.classification.OTXMulticlassClsLitModule",
+    OTXTaskType.MULTI_LABEL_CLS: "otx.core.model.module.classification.OTXMultilabelClsLitModule",
+    OTXTaskType.H_LABEL_CLS: "otx.core.model.module.classification.OTXHlabelClsLitModule",
+    OTXTaskType.DETECTION: "otx.core.model.module.detection.OTXDetectionLitModule",
+    OTXTaskType.INSTANCE_SEGMENTATION: "otx.core.model.module.instance_segmentation.OTXInstanceSegLitModule",
+    OTXTaskType.SEMANTIC_SEGMENTATION: "otx.core.model.module.segmentation.OTXSegmentationLitModule",
+    OTXTaskType.ACTION_CLASSIFICATION: "otx.core.model.module.action_classification.OTXActionClsLitModule",
+    OTXTaskType.ACTION_DETECTION: "otx.core.model.module.action_detection.OTXActionDetLitModule",
+}
+
+
+class Engine:
+    """OTX Engine.
+
+    This class defines the common interface for OTX, including methods for training and testing.
+
+    Example:
+        # Auto-Configuration with data_root
+        >>> engine = Engine(
+            data_root=<dataset/path>,
+        )
+        # With Custom OTXModel
+        >>> engine = Engine(
+            data_root=<dataset/path>,
+            model=OTXModel(...),
+            checkpoint=<checkpoint/path>,
+        )
+        # With Custom OTXDataModule
+        >>> engine = Engine(
+            model = OTXModel(...),
+            datamodule = OTXDataModule(...),
+        )
+    """
+
+    def __init__(
+        self,
+        *,
+        data_root: str | Path | None = None,
+        task: OTXTaskType | None = None,
+        work_dir: str | Path = "./otx-workspace",
+        datamodule: OTXDataModule | None = None,
+        model: OTXModel | str | None = None,
+        optimizer: OptimizerCallable | None = None,
+        scheduler: LRSchedulerCallable | None = None,
+        checkpoint: str | None = None,
```

Comment:
As I understand it, the current metric is held by the LightningModule for each task, so I don't know the proper way to change it, so this is something that needs to be discussed.

Reply:
Yeah, I found it out in #2788, but not sure if this is a good idea. It is good to have a default metric for each task. However, it is not good to hardcode the metrics for each task. A user might just want to try another metric for the task. For example, it `Accuracy` for the classification task, but I might just want to check the `F1Score` as well.

Reply:
Since the base module is designed this way, maybe we could leave it for another PR, but I believe we should discuss this. Can you add this to your agenda to discuss please?

Reply:
Sure. I'd like to discuss that agenda in the Forum tomorrow rather than in the current PR, and hopefully it can be addressed in another PR in the future.